UNITTESTS/drivers/PwmOut/test_pwmout.cpp

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (c) 2019 Arm Limited and affiliates.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "gtest/gtest.h"
+#include "gmock/gmock.h"
+#include "drivers/PwmOut.h"
+
+using namespace mbed;
+
+class MbedPowerMgmtInterface {
+public:
+    virtual ~MbedPowerMgmtInterface() {}
+    virtual void sleep_manager_unlock_deep_sleep_internal() = 0;
+    virtual void sleep_manager_lock_deep_sleep_internal() = 0;
+};
+
+
+class MockMbedPowerMgmt : public MbedPowerMgmtInterface {
+public:
+    virtual ~MockMbedPowerMgmt() {}
+    MOCK_METHOD0(sleep_manager_unlock_deep_sleep_internal, void());
+    MOCK_METHOD0(sleep_manager_lock_deep_sleep_internal, void());
+
+    static MockMbedPowerMgmt *get_instance(bool create_instance)
```

Comment:
Because it is good coding practice to have one when implementing a singleton class with a method to release the instance. As one does not want to release an instance in a singleton if it is still referenced.

Reply:
This function should return a pointer to the static instance i.e. it should not be possible to return `NULL`. If `create_instance` is false, the function can return NULL. Also if you are keeping a reference count I don't see why this is not matching in `release_instance`.

Reply:
Because of what I mentioned [here](https://github.com/ARMmbed/mbed-os/pull/11641#discussion_r340864746), I specifically want it to return a null pointer if the call came from instantiating the `PwmOut` object in `Setup` so the mock object expectation count does not increase.

I don't understand what you mean by `..why this not matching in release_instance`.  `release_instance` decrements the counter and checks that it is zero before delete the instance.

Reply:
In a singleton pattern, it is expected that a call to get_instance() does not return NULL as it returns a pointer to a static instance. So this is counterintuitive.

In addition, this extra complexity (passing a `create_instance` argument and checking references) does not provide value to the code under test; at best it would ensure that the test framework is not leaking memory but it is not able to do so because  the instance is deleted in the teardown call.

> Also if you are keeping a reference count I don't see why this is not matching in release_instance.
Sorry this is extremely badly worded. I meant to say that some calls to `get_instance` increment `ref_counter` but all calls to `release_instance` decrement the counter. Again, I fell this is counterintuitive.

Reply:
The following force-pushes address what is mentioned in your last comment.
*  https://github.com/ARMmbed/mbed-os/compare/db1c269192cd81824596368ae197007ff4266d4b..ab2544b04d2ff54af0c9ae8d26bad2f5028c7132
*  https://github.com/ARMmbed/mbed-os/compare/ab2544b04d2ff54af0c9ae8d26bad2f5028c7132..8bcddc487357de00471d299ea00842fa12506638